File: packages/core/src/db/migrate.ts
/**
 * Migration runner.
 *
 * Reads numbered `.sql` files from a migrations directory and applies them
 * in filename order, tracking each applied migration in a `_migrations`
 * table so that subsequent runs are no-ops for already-applied files.
 *
 * Each migration file is executed inside its own transaction. If execution
 * fails the transaction is rolled back and the error is re-thrown, leaving
 * the database in the state it was in before that migration started.
 */

import { Database } from "bun:sqlite";
import { readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { migrateHeadersToSummaries } from './migrations/006_summaries_data'

/**
 * Check whether the `_migrations` table already exists.
 */
function migrationsTableExists(db: Database): boolean {
  const row = db
    .query(
      "SELECT name FROM sqlite_master WHERE type = 'table' AND name = '_migrations'",
    )
    .get() as { name: string } | null;
  return row !== null;
}

/**
 * Return the set of migration names that have already been recorded.
 *
 * Migration names are stored WITHOUT the `.sql` extension (e.g.
 * `001_initial`, not `001_initial.sql`), matching the convention used
 * inside the migration files themselves.
 */
function getAppliedMigrations(db: Database): Set<string> {
  if (!migrationsTableExists(db)) {
    return new Set();
  }
  const rows = db.query("SELECT name FROM _migrations").all() as {
    name: string;
  }[];
  return new Set(rows.map((r) => r.name));
}

/**
 * Derive the migration name from a filename by stripping the `.sql` suffix.
 */
function migrationName(filename: string): string {
  return filename.replace(/\.sql$/, "");
}

/**
 * Apply all unapplied `.sql` migrations found in `migrationsDir`.
 *
 * Migrations are sorted lexicographically by filename (the three-digit
 * zero-padded prefix ensures correct ordering). Each file is executed in
 * a single transaction so that a failure in one migration does not leave
 * the database in a partially-applied state.
 *
 * The `_migrations` table is expected to be created by the initial
 * migration file itself (001_initial.sql). This runner does NOT create it
 * ahead of time, because the migration file uses `CREATE TABLE` (without
 * `IF NOT EXISTS`) and would conflict.
 */
export function runMigrations(db: Database, migrationsDir: string): void {
  // Discover migration files and sort by name.
  const files = readdirSync(migrationsDir)
    .filter((f) => f.endsWith(".sql"))
    .sort();

  const applied = getAppliedMigrations(db);

  const pending = files.filter((f) => !applied.has(migrationName(f)));

  if (pending.length === 0) {
    console.log("All migrations up to date");
    return;
  }

  for (const file of pending) {
    const sql = readFileSync(join(migrationsDir, file), "utf-8");
    const name = migrationName(file);

    // Table-rebuild migrations need PRAGMA foreign_keys = OFF, which must be
    // set OUTSIDE any transaction (SQLite silently ignores it inside BEGIN).
    // Detect these migrations by their PRAGMA statement and handle accordingly.
    const needsFkOff = sql.includes("PRAGMA foreign_keys = OFF");
    if (needsFkOff) {
      db.exec("PRAGMA foreign_keys = OFF");
    }

    try {
      db.exec("BEGIN");
      db.exec(sql);

      // Record the migration. The SQL file may already insert its own
      // record (as 001_initial.sql does), so use INSERT OR IGNORE to
      // avoid a UNIQUE constraint failure on the name column.
      db.run("INSERT OR IGNORE INTO _migrations (name) VALUES (?)", [name]);

      db.exec("COMMIT");
      if (needsFkOff) {
        db.exec("PRAGMA foreign_keys = ON");
      }
      console.log(`Applied migration: ${file}`);

      // Run companion TypeScript data migrations
      if (name === '006_summaries') {
        try {
          const result = migrateHeadersToSummaries(db)
          console.log(`  Data migration: ${result.migrated} summaries created, ${result.skipped} resumes skipped`)
        } catch (err) {
          console.error(`  Data migration failed for ${name}:`, err)
          throw err
        }
      }
    } catch (err) {
      try {
        db.exec("ROLLBACK");
      } catch {
        // Rollback may fail if the transaction was already aborted by SQLite.
      }
      if (needsFkOff) {
        db.exec("PRAGMA foreign_keys = ON");
      }
      console.error(`Migration failed: ${file}`);
      throw err;
    }
  }
}

File: crates/forge-wasm/src/migrate.rs
//! WASM migration runner — applies the shared `forge_core::migrations::MIGRATIONS`
//! slice through wa-sqlite. Mirrors `forge_sdk::db::migrate::run_migrations`
//! semantics:
//!
//! - Each migration runs in its own transaction.
//! - Migrations containing `PRAGMA foreign_keys = OFF` get the pragma
//!   set outside the transaction (SQLite ignores it inside BEGIN/COMMIT).
//! - Applied state persists in a `_migrations(name TEXT PRIMARY KEY)` table.

use forge_core::ForgeError;
use forge_core::migrations::MIGRATIONS;

use crate::database::{Database, StepResult};

/// Apply all pending migrations. Returns the count of newly-applied entries.
pub async fn run_migrations(db: &Database) -> Result<usize, ForgeError> {
    ensure_migrations_table(db).await?;
    let applied = get_applied(db).await?;
    let mut count = 0;

    for &(name, sql) in MIGRATIONS {
        if applied.iter().any(|n| n == name) {
            continue;
        }

        let needs_fk_off = sql.contains("PRAGMA foreign_keys = OFF");
        if needs_fk_off {
            db.exec_batch_internal("PRAGMA foreign_keys = OFF").await?;
        }

        db.exec_batch_internal("BEGIN").await?;
        let apply = async {
            db.exec_batch_internal(sql).await?;
            // Insert via prepared statement to bind the migration name safely.
            let stmt = db.prepare("INSERT OR IGNORE INTO _migrations (name) VALUES (?1)").await?;
            stmt.bind_text(1, name)?;
            match stmt.step().await? {
                StepResult::Done | StepResult::Row => {}
            }
            stmt.finalize().await?;
            Ok::<(), ForgeError>(())
        };

        match apply.await {
            Ok(()) => {
                db.exec_batch_internal("COMMIT").await?;
            }
            Err(e) => {
                let _ = db.exec_batch_internal("ROLLBACK").await;
                if needs_fk_off {
                    let _ = db.exec_batch_internal("PRAGMA foreign_keys = ON").await;
                }
                return Err(e);
            }
        }

        if needs_fk_off {
            db.exec_batch_internal("PRAGMA foreign_keys = ON").await?;
        }
        count += 1;
    }

    Ok(count)
}

/// Ensure the `_migrations` table exists. Idempotent.
async fn ensure_migrations_table(db: &Database) -> Result<(), ForgeError> {
    db.exec_batch_internal(
        "CREATE TABLE IF NOT EXISTS _migrations (\
            name TEXT PRIMARY KEY, \
            applied_at TEXT DEFAULT (datetime('now'))\
         )",
    ).await
}

/// Read applied migration names from `_migrations`.
async fn get_applied(db: &Database) -> Result<Vec<String>, ForgeError> {
    let stmt = db.prepare("SELECT name FROM _migrations").await?;
    let mut names = Vec::new();
    loop {
        match stmt.step().await? {
            StepResult::Row => {
                if let Some(name) = stmt.column_text(0) {
                    names.push(name);
                }
            }
            StepResult::Done => break,
        }
    }
    stmt.finalize().await?;
    Ok(names)
}
